src/api/integration/transactions_test.go

```diff
@@ -253,35 +288,314 @@ func TestLiveInjectTransactionEnableNetworking(t *testing.T) {
+				txn.InnerHash = txn.HashInner()
+				// Sign txn again as the inner hash is changed
+				txn = reSignTxnFunc(t, txn, txnRsp, w)
+
+				return &txn
+			},
+			code: 400,
+			err:  "400 Bad Request - Transaction violates hard constraint: Insufficient coins",
+		},
+		{
+			name: "transaction may not destry coins",
+			createTxn: func(t *testing.T) *coin.Transaction {
+				txn, txnRsp := prepareTxnFunc(t, w.Entries[0].Address.String(), totalCoins, "1")
+				// Make output coins < input coins
+				txn.Out[0].Coins = txn.Out[0].Coins - 1
+				txn.InnerHash = txn.HashInner()
+				// Sign the txn again as the inner hash is changed
+				txn = reSignTxnFunc(t, txn, txnRsp, w)
+				return &txn
+			},
+			code: 400,
+			err:  "400 Bad Request - Transaction violates hard constraint: Transactions may not destroy coins",
+		},
+		{
+			name: "insufficient coin hours",
+			createTxn: func(t *testing.T) *coin.Transaction {
+				txn, txnRsp := prepareTxnFunc(t, w.Entries[0].Address.String(), totalCoins, "1")
+				// Make up more output coin hours
+				txn.Out[0].Hours = txn.Out[0].Hours + 1e6
+				// Recalculate inner hash
+				txn.InnerHash = txn.HashInner()
+				// Sign the txn again
+				txn = reSignTxnFunc(t, txn, txnRsp, w)
+				return &txn
+			},
+			code: 400,
+			err:  "400 Bad Request - Transaction violates hard constraint: Insufficient coin hours",
+		},
+		{
+			name: "invalid amount, too many decimal places",
+			createTxn: func(t *testing.T) *coin.Transaction {
+				// Make a txn with txn.Out[0].Coins equal 1e3, as we have at least 2e6 coins
+				// so there will have at least two outputs.
+				txn, txnRsp := prepareTxnFunc(t, w.Entries[0].Address.String(), 1e3, "1")
+				// Make txn.Out[0].Coins too many decimal places
+				txn.Out[0].Coins = 5e2
+				// Move the remaining 5e2 from the first output to the second output, so that
+				// we won't lose coins
+				txn.Out[1].Coins = txn.Out[1].Coins + 5e2
+				txn.InnerHash = txn.HashInner()
+
+				txn = reSignTxnFunc(t, txn, txnRsp, w)
+				return &txn
+			},
+			code: 400,
+			err:  "400 Bad Request - Transaction violates soft constraint: invalid amount, too many decimal places",
 		},
 	}
+	// TODO:
+	// The following test cases can not be added here, as we cannot use the big size
+	// transaction is not allowed in encoder/decoder, which will fail the test
+	// in building transaction step.
+	//
+	// 1. Make up a txn which exceeds max block size to violate the soft constraint
+	// 2. Make up a transaction that can exceed max block size
+	// 		expected err: "400 Bad Request - Transaction violates hard constraint: Transaction size bigger than max block size"
+	// 3. Make up a transaction that has inputs/outputs exceed max
+	//		expected err:
+	// 		- "400 Bad Request - Transaction violates hard constraint: Too many signatures and inputs"
+	// 		- "400 Bad Request - Transaction violates hard constraint: Too many outputs"
+	// TODO:
+	// 1. Add test case to inject transaction who has inputs locked.
```

Comment:
I don't know if this can be tested without a very different test setup (blockchain from scratch).